src/types/query.d.ts

```diff
@@ -173,29 +315,66 @@ export module query {
   export function Year(expr: ExprArg): Expr
   export function Month(expr: ExprArg): Expr
 
+  // TODO
   export function NextId(): Expr
   export function NewId(): Expr
   export function Database(name: ExprArg, scope?: ExprArg): Expr
-  export function Index(name: ExprArg, scope?: ExprArg): Expr
-  export function Class(name: ExprArg, scope?: ExprArg): Expr
-  export function Collection(name: ExprArg, scope?: ExprArg): Expr
-  export function Function(name: ExprArg, scope?: ExprArg): Expr
+
+  // TODO: "scope" argument
+  export function Index<T extends object>(
+    name: ExprVal<string>,
+    scope?: ExprArg
+  ): Expr.IndexRef<T>
+
+  // TODO: "scope" argument
+  export function Collection<T extends object>(
+    name: ExprVal<string>,
+    scope?: ExprArg
+  ): Expr.CollectionRef<T>
+
+  // TODO: "scope" argument
+  export function Function<T>(
+    name: ExprVal<string>,
+    scope?: ExprArg
+  ): Expr.FunctionRef<T>
+
+  // TODO
   export function Role(name: ExprArg, scope?: ExprArg): Expr
   export function Databases(scope?: ExprArg): Expr
-  export function Classes(scope?: ExprArg): Expr
-  export function Collections(scope?: ExprArg): Expr
-  export function Indexes(scope?: ExprArg): Expr
-  export function Functions(scope?: ExprArg): Expr
+
+  // TODO: "scope" argument
+  export function Collections(scope?: ExprArg): Expr.SetRef<Collection<any>>
```

Comment:
it will return `Expr.SetRef<CollectionRef<any>>`

Reply:
right. it currently does do that ;)